Faster bounding box for CellReference

tvt173 opened the pull request:
The current implementation of `get_bounding_box()` for `CellReference` becomes extremely inefficient, both in time and memory, for large hierarchical cells, because it calls `self.get_polygons(depth=None)`, which will calculate the full flattened set of shapes for a reference. For large hierarchical designs, we have found multiple issues where (1) this crashes 32 bit machines (2) this becomes the major bottleneck in execution time for PIC creation.

There is no reason why we must get the flattened set of shapes here. We only need to get the bounding box of the `ref_cell` and apply the proper transformations to it, consistent with how they are done elsewhere for `CellReference`. This PR
1. Refactors out the transformation of polygons for a `CellReference` into an internal method, `_transform_polygons()`
2. Updates `CellReference.get_polygons()` to use this new method (no change to performace)
3. Uses this new method within an updated `get_bounding_box()` method, as described above, dramatically increasing efficiency

tvt173 commented:
sorry, should have run tests first... everything looks fixed now though, @heitzmann, except that there seems to be an error regarding python 3.5 reaching EOL. please consider merging

heitzmann commented:
Thanks for the PR. Transforming the bounding box does not work for arbitrary rotations, though. Only multiples of 90°, so those should be a special case. In the general case, the polygons themselves must be used for an accurate bounding box.

heitzmann commented:
Another issue that I just remembered: deeper hierarchies might be a problem. Consider the following:
```python
import gdspy

def assert_bb(bb1, bb2):
     assert all(abs(c1 - c2) < 1e-12 for p1, p2 in zip(bb1, bb2) for c1, c2 in zip(p1, p2))

cell1 = gdspy.Cell("1")
cell1.add(gdspy.Rectangle((0, 0), (1, 1)))

cell2 = gdspy.Cell("2")
cell2.add(gdspy.CellReference(cell1, rotation=45))

assert_bb(cell2.get_bounding_box(), ((-1 / 2 ** 0.5, 0), (1 / 2 ** 0.5, 2 ** 0.5)))

cell3 = gdspy.Cell("3")
cell3.add(gdspy.CellReference(cell2, rotation=45))

assert_bb(cell3.get_bounding_box(), ((-1, 0), (0, 1)))

cell4 = gdspy.Cell("4")
cell4.add(gdspy.CellReference(cell2))

assert_bb(cell4.get_bounding_box(), ((-1 / 2 ** 0.5, 0), (1 / 2 ** 0.5, 2 ** 0.5)))

cell5 = gdspy.Cell("5")
cell5.add(gdspy.CellReference(cell4, rotation=45))

assert_bb(cell5.get_bounding_box(), ((-1, 0), (0, 1)))
```

I don't think there is an easy way to get these result correctly by transforming bonding boxes. What do you think?

tvt173 commented:
@heitzmann , good points! I overlooked references with non-cardinal rotations. In this case, I suggest we fall back to calculating based on the polygons, as you point out. please see my updated PR.

I added your suggested tests firsts and confirmed
1. They pass on the current master
2. The failed on what I originally proposed

Then I made an edit, as you'll see, that falls back to flattening polygons in the reference in the case that `rotation is not None and rotation % 90 != 0`.

I think this is a pretty fair and safe solution that still dramatically speeds up the majority of cases, do you agree?

tvt173 commented:
It's also maybe worth noting that the previous solution will only ever **_overestimate_** the bounding box, which still serves the purpose of bounding_box for most applications, and may even sometimes be the preferred result. Consider what happens in the original solution for the non-90 rotation... you will effectively get the bounding box as shown below
![image](https://user-images.githubusercontent.com/3070522/102310097-a11f3080-3fad-11eb-92fe-3701955409cb.png)

Where the black box 'MyCell' indicates the bounding box of the ref_cell, and the blue box represents the bounding box which will be reported for the reference.

For the current gdspy implementation, you get the blue box below instead, which  will actually _not_ fit the bounds of the ref_cell's bounding box. So in some sense, technically it does _not_ bound every object within the cell.
![image](https://user-images.githubusercontent.com/3070522/102311209-c9a82a00-3faf-11eb-887f-bfa5aedbd0d7.png)


Which is _better_ is maybe dependent on how this is used, but maybe it is best simply not to cause a regression, though we do take a bit of a performance hit. But I would argue that the case of a non-90 degree rotation is typically rare, and more common for lower-level cells, which will incur less of a performance hit anyways than a larger, deeper, top level cell by flattening the polygons.

heitzmann commented on gdspy/library.py:
```diff
@@ -1252,9 +1252,9 @@ def area(self, by_spec=False):
+        Transforms a set of polygons based on the CellReference's applied transformations.
 
         Parameters
         ----------
```
Missing documentation of `polygons`. It should be similar to the output of `get_polygons`.
The argument `by_spec` should be removed. It is implicit by checking whether `isinstance(polygons, dict)`.

tvt173 replied:
fixed

heitzmann commented:
I agree that it is definitely better to operate on the reference bonding box for rotations multiple of 90°. Your implementation looks good to me. Can you fix the comments about the arguments in `_transform_polygons` before merging?
Also, if you could make the same changes in `CellArray`, I believe the same optimization applies. Otherwise I can do it after merging.

tvt173 commented:
hi @heitzmann , I removed the by_spec argument from `_transform_polygons()` and fixed the documentation. can we maybe merge this first and do `CellArray` as a follow-up?

heitzmann commented:
Sounds good! Thanks again!

tvt173 commented:
you are welcome! thank **you** for developing and sharing this great project! if you could please ping me when this update makes it into the pip and conda releases, that would be greatly appreciated. i'm also interested to try out your new gdstk package. what are your thoughts on its stability/ production-readiness?

heitzmann commented:
I've just published the new version to pypi. Should be on conda soon.

I consider Gdstk still unstable in terms of API, i.e., there will be breaking changes between versions until we reach 1.0. I expect those to be minor, but still problematic if you're using it on a large scale, I imagine. I try to minimize the differences between Gdstk and Gdspy when performance is not an issue, so the Python API is more stable than the C++ API; so much that I haven't even bothered to start documenting the C++ side. Right now I'm trying to add support for Oasis files (if that is interesting to you), so big changes are coming.

amccaugh commented:
Just to chime in here, there is a way to keep the bounding box tight even under non-cardinal rotations, but it  requires computing/caching the [convex hull](https://en.wikipedia.org/wiki/Convex_hull) (see [this example](http://dwoll.de/rexrepos/posts/diagBounding.html)).  However, the Clipper library doesn't provide this functionality and it's not clear to me that's going to be much of a performance improvement if it's written in pure Python, so maybe it's something better left to gdstk

heitzmann commented:
It seems that [qhull](https://github.com/qhull/qhull) might be a good choice to use, because their license seem very open. CGAL's convex hull package is GPL, so incompatible with BSL (from my understanding).

In any case, I've opened [an issue](https://github.com/heitzmann/gdstk/issues/22) in the Gdstk repository if you want to continue the discussion there.

tvt173 commented:
Hi @amccaugh , thanks for chiming in. I also thought about this a bit. I think there's one other option here too that may be a bit faster than a convex hull...
If you look at my diagrams above, one thing you'll notice is that the current implementation we apply for 90 degree rotations will _always_ give you an _over_-approximation of the bounding box for the reference, even at odd angles. If you take the bounding box and translate it back into the reference frame of the ref_cell, then all you need to do is find the closest point in the cell to each side of the over-approximated bounding box. Then, collapse each side of the box inwards to be coincident with it's closest point in the cell. That should form the true bounding box.
You still have the same issue, however, about deep references, as @heitzmann mentioned. But you could test their bounding boxes first as an approximation and only peak in and expand if one of the points of the bounding box registers closest to any of the sides.
So it still becomes, I think, a bit more complicated in either case. And this is why I question if it's worth it... I don't see many practical cases where a large hierarchical cell is rotated by an arbitrary angle... which is the only case that would cause the current fallback method (using `ref.get_polygons()`) to be critically slow or memory-intensive.
